packages/core/saved-objects/core-saved-objects-migration-server-internal/src/core/build_active_mappings.ts

```diff
@@ -57,84 +59,82 @@ export function diffMappings(actual: IndexMapping, expected: IndexMapping) {
+    (type) => {
+      const indexHashOrVersion = actual._meta!.migrationMappingPropertyHashes![type];
+      const appVersion = expected._meta!.migrationMappingPropertyHashes![type];
+      return isTypeUpdated({ type, indexHashOrVersion, appVersion, hashToVersionMap });
+    }
   );
 
-  return updatedHashes;
+  return updatedTypes;
 };
 
-// Convert an object to an md5 hash string, using a stable serialization (canonicalStringify)
-function md5Object(obj: any) {
-  return crypto.createHash('md5').update(canonicalStringify(obj)).digest('hex');
+interface HashToVersionMapParams {
+  type: string;
+  indexHashOrVersion: string;
+  appVersion: string;
+  hashToVersionMap: Record<string, string>;
 }
 
-// JSON.stringify is non-canonical, meaning the same object may produce slightly
-// different JSON, depending on compiler optimizations (e.g. object keys
-// are not guaranteed to be sorted). This function consistently produces the same
-// string, if passed an object of the same shape. If the outpuf of this function
-// changes from one release to another, migrations will run, so it's important
-// that this function remains stable across releases.
-function canonicalStringify(obj: any): string {
-  if (Array.isArray(obj)) {
-    return `[${obj.map(canonicalStringify)}]`;
-  }
-
-  if (!obj || typeof obj !== 'object') {
-    return JSON.stringify(obj);
-  }
-
-  const keys = Object.keys(obj);
-
-  // This is important for properly handling Date
-  if (!keys.length) {
-    return JSON.stringify(obj);
-  }
-
-  const sortedObj = keys
-    .sort((a, b) => a.localeCompare(b))
-    .map((k) => `${k}: ${canonicalStringify(obj[k])}`);
-
-  return `{${sortedObj}}`;
-}
-
-// Convert an object's values to md5 hash strings
-function md5Values(obj: any) {
-  return mapValues(obj, md5Object);
+/**
+ *
+ * @param type The saved object type to check
+ * @param indexHashOrVersion The current "level" of the saved object, stored in the mappings._meta (it can be either a hash or a version)
+ * @param appVersion The expected "level" of the saved object, according to the current Kibana version
+ * @returns True if the type has changed since Kibana was last started
+ */
+function isTypeUpdated({
+  type,
+  indexHashOrVersion,
+  appVersion,
+  hashToVersionMap,
+}: HashToVersionMapParams): boolean {
+  const indexEquivalentVersion = hashToVersionMap[`${type}|${indexHashOrVersion}`];
+  return indexHashOrVersion !== appVersion && indexEquivalentVersion !== appVersion;
 }
 
 // If something exists in actual, but is missing in expected, we don't
 // care, as it could be a disabled plugin, etc, and keeping stale stuff
 // around is better than migrating unecessesarily.
-function findChangedProp(actual: any, expected: any) {
-  return Object.keys(expected).find((k) => actual[k] !== expected[k]);
+function findChangedProp(actual: any, expected: any, hashToVersionMap: Record<string, string>) {
```

Comment:
NIT: I know this was there already, but can somehow use something more precise than `any` for those parameters?